[Feat/#238] FireBase RemoteConfig를 통한 앱 버전 관리를 구현합니다.

MoonsuKang opened the pull request:
## 📌 ISSUE
<!--이슈 번호 및 제목을 적어주세요!-->
closed #238
Firebase RemoteConfig를 통한 앱 버전 관리 및 업데이트 전략 적용

## 📄 Work Description
<!--어떤 작업을 했는지 작성해주세요!-->
- Firebase RemoteConfig 기반 앱 버전 관리 로직 도입
- RemoteConfig 설정 값 (min_required_version, latest_version) 으로 Soft / Hard Update 분기 처리
- RemoteConfig fetch 및 버전 비교를 위한 AppUpdateChecker, RemoteConfigDataSource, VersionComparator 설계
- Splash 화면 진입 시, 앱 버전에 따른 업데이트 안내 Dialog 노출
  - Soft Update: 사용자가 업데이트 여부를 선택할 수 있음
  - Hard Update: 업데이트를 하지 않으면 앱 사용 불가
  - 사용자 행동에 따른 마켓 이동 및 앱 종료 처리



## ✨ PR Point
<!--리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요!! 코드를 넣어도 됩니다!-->
Remote Config의 fetch 주기는 Default로 12h로 설정되어있는 걸 알게 되었습니다.

### 왜 AppUpdateChecker 라는 네이밍을 했는가?
Repository 라는 건 일반적으로 Domain Object를 영속성 계층에서 조회/저장/수정하는 역할이므로 우선 Repository는 아니고, 
Policy Evaluation에 가깝기 때문에 AppUpdateChecker, AppUpdatePolicy를 고민했는데 뭔가 AppUpdateChecker가 더 직관적이여서 선택했습니다.

### 업데이트 전략
<img width="500" src="https://github.com/user-attachments/assets/45a5bbc2-cf33-45a6-8798-1452037a77e3" />

### 파이어베이스 설정
![image](https://github.com/user-attachments/assets/857a4de5-ba88-4beb-a473-aeb64ca58f41)


## 📸 ScreenShot/Video
<!--스크린샷이나 영상을 첨부해주세요! 생략 하셔도 무방합니다!-->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced the app update experience with dynamic prompts for optional and mandatory updates.
  - Introduced dialogs that guide users seamlessly to the app store when an update is available.

- **Chores**
  - Updated the app version to 1.0.7.
  - Improved remote update configuration for smoother update management.
  - Added Firebase Config KTX library for better integration of Firebase features.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MoonsuKang committed:
[ADD/#238] Firebase Config KTX 의존성 추가

MoonsuKang committed:
[FEAT/#238] 버전 비교 기능 구현

MoonsuKang committed:
[ADD/#238] 앱 업데이트 유틸리티 클래스 추가

MoonsuKang committed:
[ADD/#238] 앱 업데이트 상태를 나타내는 sealed interface 추가

MoonsuKang committed:
[ADD/#238] Firebase Remote Config 데이터 소스 클래스 추가

MoonsuKang committed:
[FEAT/#238] 앱 업데이트 상태를 확인하는 인터페이스 및 구현체 추가

MoonsuKang committed:
[ADD/#238] Firebase Remote Config 및 앱 업데이트 체크를 위한 모듈 추가

MoonsuKang committed:
[FEAT/#238] 앱 업데이트 상태 확인 기능 추가 및 초기화 로직 구현

MoonsuKang committed:
FEAT/#238] 앱 업데이트 상태에 따른 다이얼로그 추가 및 처리 로직 구현

SYAAINN commented on app/src/main/java/com/sopt/clody/presentation/utils/appupdate/AppUpdateUtils.kt:
```diff
@@ -0,0 +1,38 @@
+
+object AppUpdateUtils {
+
+    /**
+     * 마켓 이동
+     * @param context Context
+     */
```
[3] 요즘 코드 작성 간에 함수 위에 이렇게 기능과 param에 대한 comment를 남겨놓으시더라구요! 이때까지 솝트에서 프로젝트를 하다보면 리뷰 간에 주석, 로그 제거에 좀 집착(?)하던 경향이 있었는데 어떻게 생각하시나요? 지금정도의 주석은 필요한 주석이다?!

사실 리뷰를 하는 입장에서 제 개인적으로는 좋다고 생각하기는한데 주변에 현직자들의 의견이나 문수님 의견 들은 바가 있다면 공유받고 싶습니다😀

MoonsuKang replied:
좋은 질문입니다!

## 주석 관련 생각?

사실 이 부분은 회사마다 다소 지향하는 바가 다를 수 있는데, 공식적인 KDoc 형식의 주석은 실무에서도 많이 사용하고 권장하는 경우가 많아요. 특히 SDK, Library, Utility Class처럼 다른 모듈이나 팀에서 재사용될 가능성이 있는 클래스나 함수에는 기능과 파라미터 설명을 남기는 게 좋습니다.

BUT, 너무 obvious한 주석(`// 클릭 시 이동`)이나 개발 도중 찍은 `Log`는 커밋 전에 정리하는 게 보통 맞습니다.

저도 개인적인 생각으로는 이런 정도의 KDoc 주석은 남기는 게 좋다고 생각하고, 특히 유틸리티성 함수들은 이런 주석이 없으면 오히려 의도가 잘 안 보일 때가 있어서 코드 리뷰할 때도 도움 되는 것 같습니다.

ps: 카카오뱅크에 재직중인 익명의 시니어 분에게 여쭈어 봤는데 이런 형태의 주석을 많이 사용한다고 해요~
[주석관련 아티클](https://adjh54.tistory.com/386)

## 그리고 왜 로그를 정리하라고 할까요?

코드리뷰를 받을 때 항상 "불필요한 로그 지워주세요~" 라는 걸 많이 받잖아요? 크게 2가지 이유가 있어요.

#### 첫 번째 - 성능 문제
로그 출력은 단순한 텍스트 출력이 아니라 내부적으로 I/O가 발생해요.
특히 애니메이션, 스크롤, 게임, 실시간 처리가 많은 구간에서 디버그 로그가 남아있으면 프레임 드랍, 렌더링 지연의 원인이 될 수 있답니다?

#### 두 번째 - 불필요한 메모리 할당 및 GC(Garbage Collection) 유발
로그를 위해 생성된 문자열, 포맷팅된 파라미터 등이 내부적으로 객체 할당을 발생시켜요ㅠ
이로 인해서 GC의 빈도가 늘어나고, 결국 앱의 퍼포먼스 저하로 이어질 수 있답니다!

※ 물론 Timber와 같은 로깅 라이브러리를 활용하면
릴리즈 빌드에서 로그를 비활성화할 수 있지만,
디버그 빌드에서도 의미 없는 로그는 노이즈가 될 수 있기 때문에
정리하는 습관을 들이는 것이 좋습니다~!!

MoonsuKang committed:
[REFACTOR/#238] RemoteConfigDataSource 개선

- RemoteConfig key 값 companion object로 상수화
- RemoteConfig 값 미존재 시 현재 앱 버전으로 fallback 처리

SYAAINN reviewed (approved):
코드가 참 이해하기 쉽게 잘 짜여졌네요.. 
슬슬 멀티모듈의 필요성이 느껴지는 시점에 온 것 같기도 합니다 껄껄,,,

MoonsuKang committed:
[FIX/#238] minimum_version 키 이름 수정

- KEY_MINIMUM_VERSION 값을 "minimum_version" → "min_required_version" 로 변경
